main/lsp/requests/completion.cc

```diff
@@ -839,8 +939,30 @@ vector<unique_ptr<CompletionItem>> CompletionTask::getCompletionItems(LSPTypeche
 
     vector<core::NameRef> similarLocals;
     if (params.enclosingMethod.exists()) {
-        auto locals = localNamesForMethod(typechecker, params.enclosingMethod, params.queryLoc);
-        similarLocals = allSimilarLocalNames(gs, locals, params.prefix);
+        // Slyly reuse `UnresolvedIdent::Kind` to both determine what kind of
+        // thing we're going to find and to determine the search space for
+        // `fieldsForClass`.
+        //
+        // TODO: for empty prefixes, we would like to complete instance/class
+        // variables along with locals.
```

Comment:
Trying to write tests for this was kind of hard.  This makes me believe it should not be that hard:

https://github.com/sorbet/sorbet/blob/master/main/lsp/requests/completion.cc#L1074-L1081

But writing an experimental test:

```ruby
  sig {void}
  def assign_method
    x = 
    #   ^ completion: (nothing)
    y = 5
  end
```

indicates that Sorbet thinks there are a lot of things to complete here:

```
test/testdata/lsp/completion/instance_var_strict.rb:28: ERROR: Expected completion contents:
28:     x =
           ^ completion: (nothing)
Found completion contents:
28:     x =
           ^ completion: x, y, assign_method, more_method, other_method, some_method, CSV, class, itself, object_id, singleton_class, then, yield_self, !~, ===, =~, Array, Array, BigDecimal, Complex, Float, Hash, Integer, Pathname, Rational, String, URI, __callee__, __dir__, __method__, `, abort, at_exit, autoload, autoload?, binding, block_given?, caller, caller, caller, caller_locations, caller_locations, catch, clone, define_singleton_method, define_singleton_method, display, dup, enum_for, enum_for, eql?, eval, exec, exit, exit, exit!, extend, fail, fail, fail, fail, fork, fork, format, freeze, frozen?, gem, gets, global_variables, hash, initialize_copy, inspect, instance_of?, instance_variable_defined?, instance_variable_get, instance_variable_set, instance_variables, is_a?, iterator?, kind_of?, lambda, load, local_variables, loop, loop, method, methods, nil?, open, open, p, pp, print, printf, private_methods, proc, protected_methods, public_method, public_methods, public_send, putc, puts, raise, raise, raise, raise, rand, rand, rand, rand, readline, readlines, remove_instance_variable, require, require_relative, respond_to?, select, send, send, singleton_method, singleton_methods, sleep, sleep, sprintf, srand, syscall, system, taint, tainted?, tap, test, throw, to_enum, to_enum, to_s, trap, trap, trust, undef, untaint, untrust, untrusted?, warn, !, !=, ==, __id__, __send__, equal?, instance_eval, instance_eval, instance_exec
```

and it didn't seem worth it to bake that into the test.

Reply:
You can use `...` to encode a prefix

```
# ^ completion: x, y, ...
```

Reply:
Some other things to note:

- if you use a `module` not a `class`, you won't have `Kernel` by default, which means you won't have quite so many garbage methods. There will still be some but not nearly as many.
- maybe it's fine to never suggest instance variables here? It doesn't seem like a terrible restriction to say that you have to type `@` to get instance variable suggestions.